src/v/s3/client.cc

```diff
@@ -0,0 +1,415 @@
+          cred_builder.set_priority_string("PERFORMANCE");
+          return cred_builder.set_system_trust()
+            .then([&client_cfg, &cred_builder] {
+                return cred_builder.build_reloadable_certificate_credentials()
+                  .then(
+                    [&client_cfg](
+                      ss::shared_ptr<ss::tls::certificate_credentials> creds) {
+                        client_cfg.credentials = std::move(creds);
+                        return ss::net::dns::resolve_name(
+                          client_cfg.uri(),
+                          ss::net::inet_address::family::INET);
+                    });
+            })
+            .then([&client_cfg](ss::net::inet_address addr) {
+                const uint16_t port = 443;
+                client_cfg.server_addr = ss::socket_address(addr, port);
+                return ss::make_ready_future<configuration>(client_cfg);
+            });
+      });
+}
+
+// request_creator //
+
+request_creator::request_creator(const configuration& conf)
+  : _ap(conf.uri)
+  , _sign(conf.region, conf.access_key, conf.secret_key) {}
+
+result<http::client::request_header> request_creator::make_get_object_request(
+  bucket_name const& name, object_key const& key) {
+    http::client::request_header header{};
+    // GET /{object-id} HTTP/1.1
+    // Host: {bucket-name}.s3.amazonaws.com
+    // x-amz-date:{req-datetime}
+    // Authorization:{signature}
+    // x-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+    auto host = fmt::format("{}.{}", name(), _ap());
+    auto target = fmt::format("/{}", key());
+    std::string emptysig
+      = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+    std::string agent = "redpanda.io";
+    header.method(boost::beast::http::verb::get);
+    header.target(target);
+    header.insert(boost::beast::http::field::user_agent, agent);
+    header.insert(boost::beast::http::field::host, host);
+    header.insert(boost::beast::http::field::content_length, "0");
+    header.insert(aws_header_names::x_amz_content_sha256, emptysig);
+    auto ec = _sign.sign_header(header, emptysig);
+    if (ec) {
+        return ec;
+    }
+    return header;
+}
+
+result<http::client::request_header>
+request_creator::make_unsigned_put_object_request(
+  bucket_name const& name, object_key const& key, size_t payload_size_bytes) {
+    // PUT /my-image.jpg HTTP/1.1
+    // Host: myBucket.s3.<Region>.amazonaws.com
+    // Date: Wed, 12 Oct 2009 17:50:00 GMT
+    // Authorization: authorization string
+    // Content-Type: text/plain
+    // Content-Length: 11434
+    // x-amz-meta-author: Janet
+    // Expect: 100-continue
+    // [11434 bytes of object data]
+    http::client::request_header header{};
+    auto host = fmt::format("{}.{}", name(), _ap());
+    auto target = fmt::format("/{}", key());
+    std::string sig = "UNSIGNED-PAYLOAD";
+    std::string agent = "redpanda.io";
```

Comment:
same as above. I'd put it in that shared struct you have above or in the http client itself.